### DISCUSSION TITLE
More then 6 IRQs how?

### DISCUSSION THREAD

**Discussion by duaneellissd** [Mi-V-Soft-RISC-V/miv-rv32-documentation#15] [Mi-V RV32 Soft Processors] [answered]
In the file:
```https://github.com/Mi-V-Soft-RISCV/platform/blob/main/miv_rv32_hal/miv_rv32_plic.h```
I see this:
```
Legacy RV32 soft processors are DEPRICATED.
 *        Migrate to MIV_RV32 v3.0 or later.
 ```

In the file:
```https://github.com/Mi-V-Soft-RISC-V/platform/blob/main/miv_rv32_hal/sample_fpga_design_config.h```
It states:
```
On the legacy RV32 cores, there can be up to 31 external interrupts (IRQ[30:0] pins). The legacy RV32 Soft processor external interrupts are defined in the  miv_rv32_plic.h
```
It goes on to state:
```
The MIV_RV32 soft processor has up to six optional system interrupts, MSYS_EI[n] 
 in addition to one EXT_IRQ.
 The MIV_RV32 does not have an inbuilt PLIC and all the interrupts are directly
 delivered to the processor core, hence unlike legacy RV32 cores, no interrupt
 number mapping is necessary on MIV_RV32 core.
```
So my question is: What do you do if you need more then 6 IRQS?
For example a system with 2 uarts (each uart has 4 IRQ outputs).
What about a timer? or an I2C -Your I2C Driver does not work without IRQs and a timer.

**Comment by nitindeshpande** [comment, marked as answer]
@duaneellissd 
Since the MIV_RV32 does not have a inbuilt PLIC, the idea is to use another core called MIV_ESS. The MIV_ESS contains a PLIC.

The MIV_ESS PLIC can take up to 31 interrupt inputs. The interrupts in your systems such as 4 IRQs from each of the UART can be connected there. The new MIV_RV32 has an external interrupt input pin called EXT_IRQ along with the MSYS_EI[5:0]. 
The interrupt output of MIV_ESS can be connected to the EXT_IRQ pin on the MIV_RV32. 

In addition to the PLIC, the MIV_ESS contains other peripheral submodules as well. See [MIV_ESS UG](https://www.microsemi.com/document-portal/doc_download/1246011-mi-v-ess-user-guide).

The drivers for all submodules of MIV_ESS are available [here](https://github.com/Mi-V-Soft-RISC-V/platform/tree/main/drivers/fpga_ip) and the example projects available [here](https://github.com/Mi-V-Soft-RISC-V/miv-rv32-bare-metal-examples/tree/main/driver-examples)
Sample Libero design using MIV_RV32 and MIV_ESS are available at [Eval Kit](https://github.com/Mi-V-Soft-RISC-V/PolarFire-Eval-Kit/tree/main/Libero_Projects) and [Splash kit](https://github.com/Mi-V-Soft-RISC-V/PolarFire-FPGA-Splash-Kit/tree/main/Libero_Projects).

**Comment by nitindeshpande** [comment]
Closing as this is more of a discussion item than an issue. 
Kindly open a discussion item if more clarification is required.

**Comment by duaneellissd** [comment]
@nitindeshpande - thank you this is very informative information.
